zarr/v3/abc/group.py

```diff
@@ -17,70 +19,188 @@ def info(self) -> Any:  # TODO: type this later
+        """get child"""
+        ...
+
+    @abstractmethod
+    async def group_keys(self) -> AsyncIterator[str]:
+        """iterate over child group keys"""
+        ...
+
+    @abstractmethod
+    async def groups(self) -> AsyncIterator["AsyncGroup"]:
+        """iterate over child groups"""
+        ...
+
+    @abstractmethod
+    async def array_keys(self) -> AsyncIterator[str]:
+        """iterate over child array keys"""
+        ...
+
+    @abstractmethod
+    async def arrays(self) -> AsyncIterator[AsyncArray]:
+        """iterate over child arrays"""
+        ...
+
+    @abstractmethod
+    async def tree(self, expand=False, level=None) -> Any:  # TODO: type this later
+        ...
+
+    @abstractmethod
+    async def create_group(self, name: str, **kwargs) -> "AsyncGroup":
+        ...
+
+    @abstractmethod
+    async def create_array(self, name: str, **kwargs) -> AsyncArray:
+        ...
+
+    @abstractmethod
+    async def empty(self, **kwargs) -> AsyncArray:
+        ...
+
+    @abstractmethod
+    async def zeros(self, **kwargs) -> AsyncArray:
+        ...
+
+    @abstractmethod
+    async def ones(self, **kwargs) -> AsyncArray:
+        ...
+
+    @abstractmethod
+    async def full(self, **kwargs) -> AsyncArray:
+        ...
+
+    @abstractmethod
+    async def empty_like(self, prototype: AsyncArray, **kwargs) -> AsyncArray:
+        ...
+
+    @abstractmethod
+    async def zeros_like(self, prototype: AsyncArray, **kwargs) -> AsyncArray:
+        ...
+
+    @abstractmethod
+    async def ones_like(self, prototype: AsyncArray, **kwargs) -> AsyncArray:
+        ...
+
+    @abstractmethod
+    async def full_like(self, prototype: AsyncArray, **kwargs) -> AsyncArray:
+        ...
+
+    @abstractmethod
+    async def move(self, source: str, dest: str) -> None:
+        ...
+
+    # TODO / maybes:
     # store_path (rename to path?)
-    # __enter__
-    # __exit__
-    # group_keys
-    # groups
-    # array_keys
-    # arrays
     # visit
     # visitkeys
     # visitvalues
-    # visititems
-    # tree
-    # create_group
-    # require_group
-    # create_groups
-    # require_groups
-    # create_dataset
-    # require_dataset
-    # create
-    # empty
-    # zeros
-    # ones
-    # full
-    # array
-    # empty_like
-    # zeros_like
-    # ones_like
-    # full_like
-    # move
+
+
+class SyncGroup(BaseGroup, MutableMapping):
+    @abstractproperty
+    def nchildren(self) -> int:
+        ...
+
+    @abstractproperty
+    def children(self) -> Iterator:
+        ...
+
+    @abstractmethod
+    def __contains__(self, child: str) -> bool:
+        """check if child exists"""
+        ...
+
+    @abstractmethod
+    def __getitem__(self, child: str) -> Union[SyncArray, "SyncGroup"]:
+        """get child"""
+        ...
+
+    @abstractmethod
+    def __setitem__(self, key: str, value: Union[SyncArray, "SyncGroup"]) -> None:
+        """get child"""
```

Comment:
```suggestion
```
Proposing that we do away with `Group.__setitem__` all together. It really doesn't make sense to include in the api.